src/Nethermind/Nethermind.Network.Discovery/RoutingTable/NodeTable.cs

```diff
@@ -73,42 +76,140 @@ public void RefreshNode(Node node)
+    {
+        private readonly NodeBucket[] _buckets;
+        private readonly int _bucketSize;
+        private BondedItemsEnumerator _itemEnumerator;
+        private bool _enumeratorSet;
+        private int _bucketIndex;
+        private int _count;
+
+        public ClosestNodesEnumerator(NodeBucket[] buckets, int bucketSize)
+        {
+            _buckets = buckets;
+            _bucketSize = bucketSize;
+            Current = null!;
+            _bucketIndex = -1;
+            _count = 0;
+        }
+
+        public Node Current { get; private set; }
 
-        foreach (NodeBucket nodeBucket in Buckets)
+        object IEnumerator.Current => Current;
+
+        public bool MoveNext()
         {
-            foreach (NodeBucketItem nodeBucketItem in nodeBucket.BondedItems)
+            try
             {
-                if (count < bucketSize)
+                while (_count < _bucketSize)
                 {
-                    count++;
-                    if (nodeBucketItem.Node is not null)
+                    if (!_enumeratorSet || !_itemEnumerator.MoveNext())
                     {
-                        yield return nodeBucketItem.Node;
+                        _itemEnumerator.Dispose();
+                        _bucketIndex++;
+                        if (_bucketIndex >= _buckets.Length)
+                        {
+                            return false;
+                        }
+
+                        _itemEnumerator = _buckets[_bucketIndex].BondedItems.GetEnumerator();
+                        _enumeratorSet = true;
+                        continue;
                     }
+
+                    Current = _itemEnumerator.Current.Node!;
+                    _count++;
+                    return true;
                 }
-                else
-                {
-                    yield break;
-                }
             }
+            finally
+            {
+                _itemEnumerator.Dispose();
+            }
+
+            return false;
         }
+
+        public void Reset() => throw new NotSupportedException();
+
+        public void Dispose() { }
+
+        public ClosestNodesEnumerator GetEnumerator() => this;
+
+        IEnumerator<Node> IEnumerable<Node>.GetEnumerator() => this;
+
+        IEnumerator IEnumerable.GetEnumerator() => this;
     }
 
-    public IEnumerable<Node> GetClosestNodes(byte[] nodeId)
+    public ClosestNodesFromNodeEnumerator GetClosestNodes(byte[] nodeId)
+    {
+        return GetClosestNodes(nodeId, _discoveryConfig.BucketSize);
+    }
+
+    public ClosestNodesFromNodeEnumerator GetClosestNodes(byte[] nodeId, int bucketSize)
     {
         CheckInitialization();
+        return new ClosestNodesFromNodeEnumerator(Buckets, nodeId, _nodeDistanceCalculator, Math.Min(bucketSize, _discoveryConfig.BucketSize));
+    }
+
+    public struct ClosestNodesFromNodeEnumerator : IEnumerator<Node>, IEnumerable<Node>
+    {
+        private readonly List<Node> _sortedNodes;
```

Comment:
ArrayPoolList?

Reply:
Changed to use ArrayPoolList https://github.com/NethermindEth/nethermind/pull/6967/commits/a144110a4b712760e32dabac8f9c60e2a758b9d2